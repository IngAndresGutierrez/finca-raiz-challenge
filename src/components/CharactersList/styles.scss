.cards-container {
  display: flex;
  flex-wrap: wrap;
  gap: 1.5rem;
  justify-content: center;

  @media (max-width: 768px) {
    justify-content: center;
  }
}

.title {
  font-size: 2rem;
  color: #333;
  text-align: center;
  margin-bottom: 2rem;
  text-transform: uppercase;

  &::after {
    content: '';
    display: block;
    width: 80px;
    height: 4px;
    background-color: #007bff;
    margin: 0.5rem auto;
  }
}
