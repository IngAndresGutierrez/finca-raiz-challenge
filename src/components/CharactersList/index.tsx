import { useState } from 'react'

import './styles.scss'
import Backdrop from '../commons/Backdrop'
import CharacterCard from '../CharacterCard'
import SearchCharacter from '../SearchCharacter'
import { Character } from '@/interfaces/character'
import PaginationButtons from '../PaginationButtons'
import useCharacterList from '@/hooks/useCharacterList'

const CharactersList = () => {
  const [searchTerm, setSearchTerm] = useState('')
  const [currentPage, setCurrentPage] = useState(1)
  const { characters, isLoading } = useCharacterList(currentPage, searchTerm)

  const handleNextPage = () => {
    setCurrentPage((prevPage) => prevPage + 1)
  }

  const handlePreviousPage = () => {
    if (currentPage > 1) {
      setCurrentPage((prevPage) => prevPage - 1)
    }
  }

  const handleSearch = (searchTerm: string) => {
    setSearchTerm(searchTerm)
  }

  return (
    <>
      <div>
        <h1 className='title'>Personajes de Rick and Morty</h1>

        <SearchCharacter searchTerm={searchTerm} onSearch={handleSearch} />

        <div className='cards-container'>
          {characters.map((character: Character) => (
            <CharacterCard
              key={character.id}
              name={character.name}
              image={character.image}
              status={character.status}
              species={character.species}
            />
          ))}
        </div>

        {isLoading && <Backdrop />}

        <PaginationButtons
          currentPage={currentPage}
          isLoading={isLoading}
          handlePreviousPage={handlePreviousPage}
          handleNextPage={handleNextPage}
        />
      </div>
    </>
  )
}

export default CharactersList
