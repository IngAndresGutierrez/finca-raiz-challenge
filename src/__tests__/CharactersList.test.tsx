import { fireEvent, render, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import CharactersList from '../components/CharactersList'

describe('CharactersList', () => {
  test('renders correctly', () => {
    const { container } = render(<CharactersList />)
    expect(container).toMatchSnapshot()
  })

  test('renders title correctly', () => {
    const { getByText } = render(<CharactersList />)
    expect(getByText('Personajes de Rick and Morty')).toBeInTheDocument()
  })

  test('renders character cards', async () => {
    const { getAllByTestId } = render(<CharactersList />)

    await waitFor(() => {
      const characterCards = getAllByTestId('character-card')
      expect(characterCards.length).toBeGreaterThan(0)
    })
  })

  test('renders search character input', () => {
    const { getByPlaceholderText } = render(<CharactersList />)

    const searchInput = getByPlaceholderText('Buscar por nombre')
    expect(searchInput).toBeInTheDocument()
  })

  test('calls handleSearch function with correct search term', () => {
    const { getByPlaceholderText } = render(<CharactersList />)

    const searchInput = getByPlaceholderText('Buscar por nombre')
    fireEvent.change(searchInput, { target: { value: 'Rick' } })

    expect(searchInput).toHaveValue('Rick')
  })

  test('renders previous page button', async () => {
    const { getByText } = render(<CharactersList />)

    await waitFor(() => {
      const previousPageButton = getByText('P치gina anterior')
      expect(previousPageButton).toBeInTheDocument()
    })
  })

  test('renders next page button', async () => {
    const { getByText } = render(<CharactersList />)

    await waitFor(() => {
      const nextPageButton = getByText('P치gina siguiente')
      expect(nextPageButton).toBeInTheDocument()
    })
  })

  test('renders next page button clicked', async () => {
    const { getByText, getAllByTestId } = render(<CharactersList />)

    await waitFor(() => {
      const nextPageButton = getByText('P치gina siguiente')
      fireEvent.click(nextPageButton)
      const characterCards = getAllByTestId('character-card')
      expect(characterCards.length).toBeGreaterThan(0)
    })
  })

  test('renders previous page button clicked', async () => {
    const { getByText, getAllByTestId } = render(<CharactersList />)

    await waitFor(() => {
      const previousPageButton = getByText('P치gina anterior')
      fireEvent.click(previousPageButton)
      const characterCards = getAllByTestId('character-card')
      expect(characterCards.length).toBeGreaterThan(0)
    })
  })
})
