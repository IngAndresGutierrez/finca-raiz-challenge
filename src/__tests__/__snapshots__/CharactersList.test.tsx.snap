// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharactersList renders correctly 1`] = `
<div>
  <div>
    <h1
      class="title"
    >
      Personajes de Rick and Morty
    </h1>
    <div
      class="search-character-container"
    >
      <input
        class="search-character"
        data-testid="search-character"
        placeholder="Buscar por nombre"
        type="text"
        value=""
      />
    </div>
    <div
      class="cards-container"
    />
    <div
      class="backdrop"
    />
    <div
      class="buttons-container"
    >
      <button
        class="page-button disabled loading"
        disabled=""
      >
        Cargando...
      </button>
      <button
        class="page-button loading"
        disabled=""
      >
        Cargando...
      </button>
    </div>
  </div>
</div>
`;
